sprint_07 old

Gvozdkov opened the pull request:
Я постарался исправить все пункты и я буду верить, что сегодня звезды на моей стороне) Но не совсем понимаю, как реализовать свойство hidesWhenStopped вместо ручного isHidden, возможно сказывается, что переделываю по ночам (это не оправдание) =)).  Если, я использую hidesWhenStopped у меня все равно получается ручной метод ... все, что я нашел в сети не сильно мне помогло ... либо постоянно этот индикатор работает, либо я что-то делаю не так)

Commit:
1) Созданы 2 таргета:
	- MovieQuizTests 	- MovieQuizUITests  2) Заданы идентификаторы: 	- Yes button - Yes 	- Preview Image - Poster

Commit:
Исправлены констрейнты LaunchScreen

Commit:
Создан: в UITests метод testYesButton

Commit:
Добавлен метод testNoButton в класс MovieQuizUITests, который тестирует кнопку "Нет".

Commit:
Добавлен метод testAlertsFinal в класс MovieQuizUITests, который тестирует алерт с результатом квиза и запускает игру заново.

Commit:
I. MovieQuiz ->  Services ->  StatisticService :
1.  Переменная totalAccuracy  переписана на более лаконичный вид
2.  Переменным bestGame, gamesCount присвоен статус private

II. MovieQuiz ->  Services -> NetworkClient:
1. Убран протокол NetworkRouting его реализация вошла в структуру NetworkClient

III. MovieQuiz ->  Services -> MoviesLoader:
1. Из - за реализации NetworkRouting в пункте II, в структуре MoviesLoader убран лишний инициализатор

Commit:
Пересобран MovieQuizViewController от лишнего и дублирующего кода

Commit:
add new file MovieQuizPresenter

Commit:
1. Создал файл с MovieQuizPresenter;
2. Перен туда метод convert(model: QuizQuestion);
3. Скопировал в presenter свойства questionsAmount и currentQuestionIndex;
4. Добавил свойство presenter в MovieQuizViewControler;
5. Применимл DRY для устранения дублирования данных.
6. Создан метод blockButton() в MovieQuizViewControler для дольнейшей реализации блокировки кнопок при пелеключение ответов

Commit:
Перемена часть логики из MovieQuizViewController в MovieQuizPresenter

Commit:
1. Из MovieQuizPresenter в MovieQuizViewController вернул метод showNextQuestionOrResults
2. В MovieQuizViewController убран метод  blockBatton его реализация вошла в showNextQuestionOrResults
Проект компилируется и запускается на эмуляторе.

Commit:
Перенесен метод showNextQuestionOrResults из MovieQuizViewController в MovieQuizPresenter

Commit:
Перенесено QuestionFactoryDelegate в MovieQuizViewController

Commit:
Отлично! Мы с вами окончательно вынесли всю «не View» логику из MovieQuizViewController в презентер MovieQuizPresenter.

Commit:
Создан новый таргет MovieQuizPresenterTest. Тесты успешно завершаются.

Commit:
1. Исправлена ошибка в MovieQuizPresenterTest с "Value of type 'any Movie QuizViewController Protocol' has no member 'block Buttons'".

Commit:
Проведен рефакторинг MovieQuizViewController, presenter теперь работает, как с  протоколом QuestionFactoryDelegate, а не как с классом  MovieQuizPresent

Commit:
Изменено направление слэша в алерте с финальным результатом

Commit:
Папка Presentation вынесена на уровень выше

Commit:
1. В MoviesLoaderTests структура StubNetworkClient вынесена выше тестовых функций.
2. Удалены закомментированные  мок-вопросы в QuestionFactory.

Commit:
В MovieQuizUITests добавлен тест testGameRound

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -9,192 +9,98 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
-    private var questionFactory: QuestionFactoryProtocol?        // та самая фабрика вопросов, которую мы создали. Наш                                                                   контроллер будет обращаться за вопросами именно к ней.
+//    private var presenter: MovieQuizPresenter!
+
+    private var presenter: MovieQuizPresenter!
```
Здесь необходимо использовать протокол QuestionFactoryDelegate и рекомендую указать его опциональным, а в протокол добавить отсутствующие методы, такие как makeResultsMessage() и т.д.

protos17 commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,145 @@
+            
+            let bestGame = statisticService.bestGame
+            
+            let totalPlaysCountLine = "Количество сыгранных квизов: \(statisticService.gamesCount)"
```
Можно использовать многострочный вариант строки

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -9,192 +9,98 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
     }
     
+    func showLoadingIndicator() {
+        activityIndicator.isHidden = false // говорим, что индикатор загрузки не скрыт
```
Рекомендую использовать свойство hidesWhenStopped вместо ручного управления свойством isHidden.

protos17 commented on MovieQuiz/Presentation/MovieQuizViewControllerProtocol.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+import Foundation
```
Два раза импортирована библиотека

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,102 @@
+final class ArrayTests: XCTestCase {
+    func testGetValueInRange() throws { // тест на успешное взятие элемента по индексу
+        //Given
+        let array = [1, 1, 2, 3, 5]
```
Рекомендую использовать уникальные значения, потому что array[safe: 0] и array[safe: 1] дадут одинаковый результат

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,102 @@
+    
+    func testGetValueOutOfRange() throws {
+        //Given
+        let array = [1, 1, 2, 3, 5]
```
Рекомендую тоже указать уникальные значения

protos17 commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,102 @@
+
+
+
+//override func setUpWithError() throws {
```
Лишний закомментированный код рекомендую удалять

Commit:
1. MovieQuizTests/ArrayTests - убран лишний закомментированный код
2. MovieQuiz/Presentation/MovieQuizViewControllerProtocol - убрано два раза импортированная библиотека

Commit:
1. В MovieQuizViewController, переменная presenter использует опциональный протокол QuestionFactoryDelegate.
2. Для реализации общения через протокол QuestionFactoryDelegate в него были добавлены методы:
	а) func makeResultsMessage() -> String,
	в) func restartGame(),
	с) func yesButtonClicked(),
	d) func noButtonClicked()

Commit:
Небольшая корректировка в makeResultsMessage

Commit:
MovieQuizTests/ArrayTests -> в методах testGetValueInRange() и testGetValueOutOfRange() заменены значения array[safe: 0] и array[safe: 1] на уникальные во избежания одинаковых ответов
